torch/testing/_internal/common_methods_invocations.py

```diff
@@ -17693,13 +17703,25 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
            supports_fwgrad_bwgrad=True,
            sample_inputs_func=sample_inputs_pca_lowrank,
            decorators=[skipCUDAIfNoCusolver, skipCPUIfNoLapack, with_tf32_off,
-                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=1e-03, rtol=1e-03)}),
-                                    'TestCommon', 'test_noncontiguous_samples',
-                                    device_type='cuda')],
+                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=4e-02, rtol=4e-02),
+                                                       torch.complex64: tol(atol=4e-02, rtol=4e-02)}),
+                                    'TestCommon', 'test_noncontiguous_samples'),
```

Comment:
Whats with the tolerance bumps for `float32`?

Reply:
Yeah, tolerance 0.04 looks unusually high for float32, it could indicate a bug.
@lezcano can you reproduce this locally? If so, forcing the samples to be contiguous (but valuewise the same as used in the noncontiguous testing) could provide an answer.

Reply:
The error here is in the gradients, and here we are differentiating an approximation algorithm. We don't have any guarantees that differentiating the algorithm even converges to the derivative, even less that it converges in the same number of iterations!

Reply:
Okay, but I thought these were supposed to be cosmetic cleanups that shouldn't effect numerics. So why are larger tolerances required?

Reply:
Locally, with
```
diff --git a/torch/testing/_internal/common_methods_invocations.py b/torch/testing/_internal/common_methods_invocations.py
index 21fb3b0561a..3d5ccb91ffe 100644
--- a/torch/testing/_internal/common_methods_invocations.py
+++ b/torch/testing/_internal/common_methods_invocations.py
@@ -17703,8 +17703,8 @@ op_db: List[OpInfo] = [
            supports_fwgrad_bwgrad=True,
            sample_inputs_func=sample_inputs_pca_lowrank,
            decorators=[skipCUDAIfNoCusolver, skipCPUIfNoLapack, with_tf32_off,
-                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=4e-02, rtol=4e-02),
-                                                       torch.complex64: tol(atol=4e-02, rtol=4e-02)}),
+                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=1e-06, rtol=1e-06),
+                                                       torch.complex64: tol(atol=4e-03, rtol=4e-03)}),
                                     'TestCommon', 'test_noncontiguous_samples'),
                        # FIXME This should be the following, but the toleranceOverride does not seem to do anything!
                        # DecorateInfo(toleranceOverride({torch.complex128: tol(atol=1e-04, rtol=1e-04)}),
```
`test_noncontiguous_samples_pca_lowrank` tests pass OK.

For smaller tolerances, these tests fail with:
```
Got different gradients for contiguous / non-contiguous inputs wrt input 1.
```
With a deterministic operation, such failures should basically never happen because the  operation result typically does not depend on data storage details.

However, `pca_lowrank` is not a deterministic operation as its algorithm uses `torch.randn` to find the initial approximation of a lowrank input. As a result, calling `pca_lowrank` several times with identical inputs will *not* produce identical results. This explains why different gradients are obtained when using contiguous or non-contiguous inputs (that valuewise are equal).

Reply:
The only change here is the clone. I now regenerate all samples for `M is not None`. I think that was the issue

Reply:
In an operation as complex as this, getting all the numbers pretty close, and one within 0.01, pretty much tells you that the gradient is alright. Even more so when the gradient is autogenerated and the forward pass is passing the tests...

Reply:
> The only change here is the clone. I now regenerate all samples for `M is not None`. I think that was the issue

So, it was https://github.com/pytorch/pytorch/pull/125580#discussion_r1616165627 after all. I cannot remember anymore after all these years why sample_close was required in the first place... could it be due to side-effects from autograd or a combination of things discussed above? Sounds like a fun task to find it out but OT here.